README.md
# Devel::Multithreaded::Cover

## Code coverage metrics for Perl

This module provides code coverage metrics for Perl.  Code coverage metrics
describe how thoroughly tests exercise code.  By using Devel::Multithreaded::Cover you can
discover areas of code not exercised by your tests and determine which tests to
create to increase coverage.  Code coverage can be considered an indirect
measure of quality.

Although it is still being developed, Devel::Multithreaded::Cover is now quite stable and
provides many of the features to be expected in a useful coverage tool.

Statement, branch, condition, subroutine, and pod coverage information is
reported.  Statement and subroutine coverage data should be accurate.  Branch
and condition coverage data should be mostly accurate too, although not always
what one might initially expect.  Pod coverage comes from Pod::Coverage.  If
Pod::Coverage::CountParents is available it will be used instead.  Coverage
data for other criteria are not yet collected.

The cover program can be used to generate coverage reports.  Devel::Multithreaded::Cover ships
with a number of reports including various types of HTML output, textual
reports, a report to display missing coverage in the same format as compilation
errors and a report to display coverage information within the Vim editor.

It is possible to add annotations to reports, for example you can add a column
to an HTML report showing who last changed a line, as determined by git blame.
Some annotation modules are shipped with Devel::Multithreaded::Cover and you can easily create
your own.

The gcov2perl program can be used to convert gcov files to "Devel::Multithreaded::Cover"
databases.  This allows you to display your C or XS code coverage together with
your Perl coverage, or to use any of the Devel::Multithreaded::Cover reports to display your C
coverage data.

Code coverage data are collected by replacing perl ops with functions which
count how many times the ops are executed.  These data are then mapped back to
reality using the B compiler modules.  There is also a statement profiling
facility which should not be relied on.  For proper profiling use
Devel::NYTProf.  Previous versions of Devel::Multithreaded::Cover collected coverage data by
replacing perl's runops function.  It is still possible to switch to that mode
of operation, but this now gets little testing and will probably be removed
soon.  You probably don't care about any of this.

The most appropriate mailing list on which to discuss this module would be
perl-qa.  See <http://lists.perl.org/list/perl-qa.html>.

The Devel::Multithreaded::Cover repository can be found at
<http://github.com/pjcj/Devel--Cover>.  This is also where problems should be
reported.

To get coverage for an uninstalled module:

    cover -test

or

    cover -delete
    HARNESS_PERL_SWITCHES=-MDevel::Multithreaded::Cover make test
    cover

To get coverage for an uninstalled module which uses Module::Build (0.26 or
later):

    ./Build testcover

If the module does not use the t/*.t framework:

    PERL5OPT=-MDevel::Multithreaded::Cover make test

If you want to get coverage for a program:

    perl -MDevel::Multithreaded::Cover yourprog args
    cover

To alter default values:

    perl -MDevel::Multithreaded::Cover=-db,cover_db,-coverage,statement,time yourprog args


Changes
Devel::Multithreaded::Cover history

{{$NEXT}}
 - Fix up tests with recent Math::BigInt versions (Ed J) (github 219)
 - Add progress text when processing coverage (Christian Walde) (github 221)
 - Optionally use Term::Size (Paul "LeoNerd" Evans) (github 228)
 - Fix indentation in pod (Graham Knop) (github 229)
 - Fix warn precedence (Alexandr Evstigneev) (github 231)

Release 1.31 - 18 September 2018
 - Add queue for cpancover (Joel Berger)
 - Handle circular refs (Ed J) (github 164, 216)
 - Allow for $DEVEL_MULTITHREADED_COVER_TEST_OPTS to be unset (Dick Franks) (github 217)

Release 1.30 - 14 July 2018
 - Fix incorrect merging of structure (Emiliya Boyadjieva) (github 199)
 - Fix up compiler warnings (Jens Rehsack) (github 211)
 - Test against 5.28.0, 5.29.0 and some other versions
 - Compress and then delete old cpancover data
 - Clarify some docs around ignore (Olaf Alders) (github 207)
 - Run coverage on cpp and hpp files (Jacques Germishuys) (github 203)
 - Add docs for coverage of modules on command line (Slaven Rezić) (github 204)
 - Ignore .AppleDouble files (David Cantrell) (github 200)

Release 1.29 - 15 October 2017
 - Lock using read and append mode (Stephan Loyd) (github 196)
 - Clean up lock files when running cover command

Release 1.28 - 5 October 2017
 - Fix typo in docs (Gregor Herrmann) (github 195)
 - Fix up options to cover program (github 194)

Release 1.27 - 20 September 2017
 - Test against 5.27.3
 - Add dependency on B::Debug (Ryan Voots) (github 189)
 - Remove obsolete dev and cpancover code
 - Manage multiple ignore and select options (Sullivan Beck) (github 163)
 - Write top-level html last for better sorting (Christian Walde) (github 158)
 - Restore XS coverage in cpancover (H.Merijn Brand) (github 188)
 - Allow cpancover to run fully from local directory
 - Handle missing start times in runs (Chad Granum) (github 190)
 - Add test for string eval (Sebastian Paaske Tørholm) (github 145)
 - Improve uncoverable tests and docs (Kevin Brannen) (github 167)
 - The cover program accepts multiple reports (Eden Hochbaum) (github 172)
 - Add -prefer_lib option to cover script (Kent Fredric) (github 91)
 - Fix low-level file locking (Graham Knop) (github 90)

Release 1.26 - 24 July 2017
 - Untaint $DEVEL_MULTITHREADED_COVER_OPTIONS (Alexey Sokolov) (github 187)
 - Add loose_perms option (Todd Rinaldo) (github 185)
 - Set EXTENDED_TESTING during cpancover runs
 - Improve version parsing for cpancover (Helmut Wollmersdorfer) (github 124)
 - Improve cpancover code and documentation

Release 1.25 - 11 May 2017
 - Fix C code to work with MSVC (Marc-Philip) (github 177)
 - Fix cover command using Module::Build on Win32 (sago35) (github 160)
 - Add file and line info to keep ops unique (Niko Tyni) (github 143)
 - Report coverage for all declarations (Heinz Knutzen) (github 166)
 - Fix spelling mistakes (Gregor Herrmann) (github 171)
 - Fix condition coverage for complex conditions (Heinz Knutzen) (github 165)
 - Use JSON::MaybeXS for sppen and flexibility (Olivier Mengué) (github 139)
 - Add CONTRIBUTING file (Gabor Szabó) (github 100)

Release 1.24 - 18 April 2017
 - Work with 5.25.x (op_sibling) (Dan Collins, Matthew Horsfall) (github 162)
 - Perl versions below 5.10.0 are now unsupported

Release 1.23 - 24 April 2016
 - Enhance html coverage popups (Haydn Newport) (github 156)
 - Add cpancover about page (Guillermo O. Freschi) (github 146)
 - Perl versions below 5.8.1 are now unsupported

Release 1.22 - 9 April 2016
 - Test against 5.20.3, 5.23.1 - 5.23.9, 5.22.1
 - Improve uncoverable docs (Alex Balhatchet) (github 148)
 - Improve cpancover

Release 1.21 - 20 September 2015
 - Test more versions on travis
 - Fix things up with cpancover for new docker release
 - Fix problem under 5.22 (Dick Franks) (github 140)

Release 1.20 - 6 July 2015
 - Document setting PerlSwitches for mod_perl (jpsalvesen) (github 128)

Release 1.19 - 5 July 2015
 - Get things working with 5.22.0
 - Test against 5.23.0
 - Remove dependency on Test::Warn

Release 1.18 - 6 April 2015
 - Remove dependency on CGI (use HTML::Entities instead) (Lee Johnson)
 - Fix a use of the wrong type of null constant (Zefram) (RT #103102)
 - Link to next uncovered statement in html_basic (H.Merijn Brand) (github 112)
 - Add "provides" to metadata (Ivan Wills) (github 119)

Release 1.17 - 20 September 2014
 - Get cpancover to release quality
 - Add IRC link to META
 - Test against 5.20.1, 5.21.1, 5.21.2, 5.21.3 and 5.21.4
 - Officially support 5.20.1

Release 1.16 - 17 August 2014
 - Fix segfault in 5.20.0
 - Improvements to cpancover
 - Fix vim report to work with recent versions (cono) (github 94)
 - Ignore vim swap files in git (Gábor Szabó) (github 95)
 - Improve help for cover command (Gábor Szabó) (github 96)

Release 1.15 - 1 June 2014
 - Allow -coverage default,-pod option to cover (David Cantrell) (github 89)
 - Get cpancover coverage in docker containers
 - Add CGI.pm as a prerequisite
 - Test against 5.20.0 and 5.21.0
 - Add longer delay in tests to try to appease *BSD

Release 1.14 - 2 May 2014
 - Fix test failures on OpenBSD, NetBSD and Windows

Release 1.13 - 28 April 2014
 - Test against 5.19.11
 - Fix "use 5.xyz;" on pre 5.10 perls (github 87)
 - Round percentages down in summary and reports (also fixes Windows tests)

Release 1.12 - 17 April 2014
 - Fix test failure on OSX, FreeBSD and HPUX (thanks H.Merijn Brand)

Release 1.11 - 13 April 2014
 - Tests should pass when all run in parallel (Kent Fredric) (github 69)
 - Fix problems with using Sereal as DB format

Release 1.10 - 6 April 2014
 - Test against 5.19.10
 - Add Sereal backend and use it if available
 - Support spaces in DB path (Keith Wissing) (github 81)
 - Test against 5.8 on Travis (Slaven Rezić) (github 76)
 - HTML cleanups (H.Merijn Brand)
 - Fix deep recursion with use overload (Matthew Horsfall) (github 78, 79)
 - Improve testing process

Release 1.09 - 15 March 2014
 - Add Contributors file
 - Test against 5.19.4 - 5.19.9

Makefile.PL
#!/usr/bin/perl

# Copyright 2001-2018, Paul Johnson (paul@pjcj.net)

# This software is free.  It is licensed under the same terms as Perl itself.

# The latest version of this software should be available from my homepage:
# http://www.pjcj.net

require 5.006001;

use strict;
use warnings;

use Cwd;
use ExtUtils::MakeMaker;
use File::Copy;
require Data::Dumper;

$| = 1;

my $Version  = "1.31";
my $Latest_t = "5.029007";
my $Author   = 'paul@pjcj.net';
my @perlbug  = ("perlbug", "-a", $Author,
                           "-s", "Installation of Devel::Multithreaded::Cover $Version");
my $Perlbug  = join " ", map { / / ? "'$_'" : $_ } @perlbug;
my %inc      = map { -d $_ ? (($_ eq "." ? $_ : Cwd::abs_path($_)) => 1) : () }
                   @INC;
my $dumper = Data::Dumper->new( [[ sort keys %inc ]] );
$dumper->Indent(1)->Terse(1);
$dumper->Sortkeys(1) if $dumper->can("Sortkeys");
chomp (my $inc_string = $dumper->Dump);

open I, ">lib/Devel/Multithreaded/Cover/Inc.pm"
    or die "Cannot open lib/Devel/Multithreaded/Cover/Inc.pm: $!";
print I <<"EOI";
# Copyright 2001-2018, Paul Johnson (paul\@pjcj.net)

# This software is free.  It is licensed under the same terms as Perl itself.

# The latest version of this software should be available from my homepage:
# http://www.pjcj.net

# This file was automatically generated by Makefile.PL.

package Devel::Multithreaded::Cover::Inc;

use strict;
use warnings;

our \$VERSION      = "$Version";
our \$Perl_version = '$]';
our \@Inc          = \@{ $inc_string };
chomp (our \$Perl  = <<'EOV');  # Careful with \\\\ in the path
$^X
EOV

if (\$Perl_version ne \$]) {
    print STDERR <<"EOM";

This version of Devel::Multithreaded::Cover was built with Perl version \$Perl_version.
It is now being run with Perl version \$].
Attempting to make adjustments, but you may find that some of your modules do
not have coverage data collected.  You may need to alter the +-inc, +-ignore
and +-select options.

EOM
    eval "use Cwd";
    my \%inc = map { -d \$_ ? ((\$_ eq "." ? \$_ : Cwd::abs_path(\$_)) => 1) : () }
                   \@INC;
    \@Inc = sort keys \%inc;
}

# TODO - check for threadedness, 64bits etc. ?

1
EOI
close I or die "Cannot close lib/Devel/Multithreaded/Cover/Inc.pm: $!";

print "Writing tests ........ ";

for my $d (qw( t t/e2e )) {
    unless (mkdir $d) {
        die "Cannot mkdir $d: $!" unless -d $d;
    }
}

my @tests;
opendir D, "tests" or die "Cannot opendir tests: $!";
for my $t (readdir D) {
    next unless -f "tests/$t";
    next if $t =~ /\.(pm|pl|org|bak|uncoverable|swp)$/;
    next if $t =~ /~$/;

    push @tests, $t;

    if ($t =~ /\.t/) {
        copy("tests/$t", "t/e2e/$t")
            or die "Cannot copy tests/$t to t/e2e/$t: $!";
        next
    }

    open T, ">t/e2e/a$t.t" or die "Cannot open t/e2e/a$t.t: $!";
    print T <<EOT;
#!$^X

# Copyright 2002-2018, Paul Johnson (paul\@pjcj.net)

# This software is free.  It is licensed under the same terms as Perl itself.

# The latest version of this software should be available from my homepage:
# http://www.pjcj.net

use strict;
use warnings;

use lib "./lib";
use lib "./blib/lib";
use lib "./blib/arch";
use lib "./t";

use Devel::Multithreaded::Cover::Test;

my \$test = Devel::Multithreaded::Cover::Test->new("$t");
\$test->run_test;
no warnings;
\$test  # for create_gold
EOT
    close T or die "Cannot close t/e2e/a$t.t: $!";
}
closedir D or die "Cannot closedir tests: $!";

s/^/tests\// for @tests;
push @tests, grep !/e2e/, glob "t/*/*.t";

if ($ENV{DEVEL_MULTITHREADED_COVER_NO_TESTS}) {
    # don't run tests under p5cover
    print "removing all tests with DEVEL_MULTITHREADED_COVER_NO_TESTS\n";
    system "rm -rf t/*";  # TODO portability
    @tests = ();
}

print "done\n\n";

my %checked;

sub check {
    my ($module, $text, $version) = @_;

    printf "checking for %-18s %-16s .... ",
           $module, $version ? "version $version" : "";

    {
        local $SIG{__WARN__} = sub {};
        eval "use $module";
    }
    (my $mod = $module) =~ s/::/\//g;
    if (my $m = $INC{"$mod.pm"}) {
        my $v = eval { no warnings; eval "\$${module}::VERSION" };
        printf "%-8s $m\n", $v;
        if ($version && $v < $version) {
            print "\n\n\n$text\n" unless $checked{$text}++;
            print "\n";
        }
    } else {
        print "not found";
        print "\n\n\n$text\n" unless $checked{$text}++;
        print "\n";
    }
};

my $d = <<EOM;
The coverage database can be stored in any of the Storable, JSON or Sereal
formats.  To use the Storable format, the Storable module is required.  This has
been a core module since perl-5.8.0.  To store the database in JSON format
JSON::MaybeXS can be used.  It will use the best JSON backend available, trying
first Cpanel::JSON::XS.  To use the Sereal format, the Sereal module should be
installed.  This is available on CPAN.

Sereal is used if it is available.  Otherwise JSON is used, if available.
Otherwise Storable will be used.  The Sereal and JSON formats should be portable
between all systems, whereas Storable format may not be.
EOM

check "Storable",      $d;
check "JSON::MaybeXS", $d;
check "Sereal",        $d;

check "Digest::MD5", <<EOM;
Digest::MD5 is required to check whether covered files have changed.  You can
download Digest::MD5 from CPAN.
EOM

check "Test::More", <<EOM;
Test::More is required to run the Devel::Multithreaded::Cover tests.  You can download
Test::More from CPAN.
EOM

check "Template", <<EOM, "2.00";
Template 2.00 is required to run the some HTML backends to cover and for
cpancover.  Unless you have specific requirements this should not be a problem,
but you will not be able to use these reports until you install the Template
Toolkit, available from CPAN.  In the meantime you may continue to use the rest
of Devel::Multithreaded::Cover.
EOM

my $m = <<EOM;
One of PPI::HTML 1.07 or Perl::Tidy 20060719 is required to add syntax
highlighting to some HTML backends to cover and for cpancover.  Unless you have
specific requirements this should not be a problem, but you will not be able to
use syntax highlighting in these reports until you install PPI::HTML or
Perl::Tidy, available from the CPAN.  However, the rest of Devel::Multithreaded::Cover will be
available as usual.
EOM

check "PPI::HTML",  $m, "1.07";
check "Perl::Tidy", $m, "20060719";

print "checking for Pod::Coverage      version 0.06     .... ";
my $e = <<EOM;
Pod::Coverage 0.06 is required to do pod coverage.  This will tell you how well
you have documented your modules.  Pod coverage will be unavailable until you
install this module, available from CPAN.  In the meantime, you may continue to
use the rest of Devel::Multithreaded::Cover.
EOM

eval "use Pod::Coverage";
if (my $m = $INC{"Pod/Coverage.pm"}) {
    my $v = eval { no warnings; $Pod::Coverage::VERSION };
    print $v < 0.06 ? "$v\n\n\n$e\n\n" : "$v     $m\n";
    print "checking for Pod::Coverage::CountParents         .... ";
    $e = <<EOM;
Pod::Coverage::CountParents.pm is used for Pod coverage if it is available.  We
will fall back to using Pod::Coverage.pm.  If you want to use
Pod::Coverage::CountParents.pm, just install it from CPAN.
EOM
    eval "use Pod::Coverage::CountParents";
    if (my $m = $INC{"Pod/Coverage/CountParents.pm"}) {
        my $v = eval { no warnings; $Pod::Coverage::CountParents::VERSION };
        $v ||= "    ";
        print "$v     $m\n";
    } else {
        print "not found\n\n\n$e\n\n";
    }
} else {
    print "not found\n\n$e\n";
}

check "Test::Differences", <<EOM;
Test::Differences is used to display output from failed tests.  Hopefully there
won't be any failed tests, but if there are you will get output that may not be
a model of clarity.  If you do get test failures and you fancy helping me by
debugging them, then you might like to consider installing Test::Differences.
You can download Test::Differences from CPAN.
EOM

check "Browser::Open", <<EOM;
Browser::Open is used to launch a web browser when the -launch flag is specified
with HTML report formats. You can download Browser::Open from CPAN.
EOM

check "HTML::Entities", <<EOM;
HTML::Entities is used to run the HTML reports.  If you would like to use the
HTML report, please install HTML::Entities.  Otherwise, other reports will be
available as usual.  Because HTML reports are expected, HTML::Entities has been
added to the prerequisites.
EOM

print <<EOM if $] > $Latest_t;

Devel::Multithreaded::Cover $Version has not been tested with perl $].
Testing will take place against expected output from perl $Latest_t.
You may well find failing tests.

EOM

if ($] < 5.010000) {
    my $v = $] < 5.008001 ? "1.22" : "1.23";
    print <<EOM;

================================================================================

                                   IMPORTANT
                                   ---------

Devel::Multithreaded::Cover $Version is not supported on perl $].  The last version of
Devel::Multithreaded::Cover which was supported was version $v.  This version may not work.
I have not tested it.  If it does work it will probably not be fully functional.

If you decide to use it anyway, you are on your own.  If it works at all, there
will be some constructs for which coverage will not be collected, and you may
well encounter bugs which have been fixed in subsequent versions of perl.
EOM

    print <<EOM if $^O eq "MSWin32";

And things are even worse under Windows.  You may well find random bugs of
various severities.
EOM
    print <<EOM;

If you are actually using this version of Devel::Multithreaded::Cover with perl $], please
let me know.  I don't want to know if you are just testing Devel::Multithreaded::Cover, only if
you are seriously using this version to do code coverage analysis of real code.
If I get no reports of such usage then I will remove support and delete the
workarounds for versions of perl below 5.10.0.

In order to use this version of Devel::Multithreaded::Cover with perl $] you must set the
environment variable \$DEVEL_MULTITHREADED_COVER_UNSUPPORTED

================================================================================

EOM
}

print "\n" x 3;

$ExtUtils::MakeMaker::Verbose = 0;

my $opts = {
    NAME             => "Devel::Multithreaded::Cover",
    VERSION          => $Version,
    AUTHOR           => 'Paul Johnson <paul@pjcj.net>',
    ABSTRACT_FROM    => "lib/Devel/Multithreaded/Cover.pm",
    DIR              => [],
    EXE_FILES        => [ map "bin/$_", qw( cover gcov2perl cpancover ) ],
    PERL_MALLOC_OK   => 1,
    PREREQ_PM        => {
                            "Storable"       => 0,
                            "Digest::MD5"    => 0,
                            "B::Debug"       => 0,
                            "HTML::Entities" => 3.69,
                            $ENV{DEVEL_MULTITHREADED_COVER_NO_TESTS}
                                ? ()
                                : ( "Test::More" => 0 )
                        },
    TYPEMAPS         => [ "utils/typemap" ],
    clean            => {
                            FILES => "t/e2e/* cover_db* t/e2e/*cover_db " .
                                     "README *.gcov *.out"
                        },
    dist             => { COMPRESS => "gzip --best --force" },
    test             => {
                            TESTS => $ENV{DEVEL_MULTITHREADED_COVER_NO_TESTS}
                                         ? ""
                                         : "t/*.t t/*/*.t"
                        },
    realclean        => $] < 5.008008 ?
                        {
                            FILES  => "lib/Devel/Multithreaded/Cover/Inc.pm",
                            POSTOP => "\$(RM_RF) cover_db t/e2e"
                        } :
                        { FILES => "lib/Devel/Multithreaded/Cover/Inc.pm cover_db t/e2e" },
};
# use Data::Dumper; print Dumper $opts;
WriteMakefile(%$opts);

print "\n";
print <<EOM if 0;

--------------------------------------------------------------------------------

I like to have some idea of the distribution of this software.  To that end I
would be very grateful for mail from you.  This will be used only to satisfy my
curiosity and to help me make decisions which may involve changes to the
module.

If you can run perlbug you can send me a success report with "make ok".
Failure reports with "make nok" are also appreciated.

If you have any questions or comments, mailing list details are available in the
README file, or send mail to me directly.

EOM

sub MY::postamble {
    my %tests;
    @tests{@tests} = map { (my $t = $_) =~ s/\W/_/g; "cover_db_$t" } @tests;
    my @reports = qw(
        compilation html_basic html_minimal html html_subtle sort text2 text vim
    );

    qq[
tags : pure_all
\t ctags --recurse --exclude=blib --exclude=Devel-Multithreaded-Cover-* .

README : lib/Devel/Multithreaded/Cover.pm
\t TERMCAP= COLUMNS=80 pod2text -s lib/Devel/Multithreaded/Cover.pm | \\
    \$(PERL) -n \\
    -e 'print if (/NAME/         ... /^[A-Z ]+\$\$/) =~ /^\\d+\$\$/;' \\
    -e 'print if (/SYNOPSIS/     ... /^[A-Z ]+\$\$/) =~ /^\\d+\$\$/;' \\
    -e 'print if (/DESCRIPTION/  ... /^[A-Z ]+\$\$/) =~ /^\\d+\$\$/;' \\
    > README

show_version :
\t \@echo \$(VERSION)

ppm : ppd pure_all
\t tar cf Devel-Multithreaded-Cover.tar blib
\t gzip --best --force Devel-Multithreaded-Cover.tar
\t \$(PERL) -pi.bak \\
    -e 's/(OS NAME=")[^"]*/\$\$1MSWin32/;' \\
    -e 's/(ARCHITECTURE NAME=")[^"]*/\$\$1MSWin32-x86-multi-thread/;' \\
    -e 's/(CODEBASE HREF=")[^"]*/\$\$1Devel-Multithreaded-Cover.tar.gz/;' \\
    Devel-Multithreaded-Cover.ppd

TAINT = -T
TAINT =
COVER_OPTIONS =
PERL5OPT =

_run : pure_all
\t \$(PERL) \$(TAINT) -Iblib/lib -Iblib/arch -MDevel::Multithreaded::Cover=-merge,0,`\$(PERL) -e '\$\$l = qx|grep __COVER__ \$\$ARGV[0]|; \$\$l =~ /__COVER__\\s+criteria\\s+(.*)/; (\$\$c = \$\$1 || "all") =~ s/\\s+/,/g; \$\$p = "\$\$1," if \$\$l =~ /__COVER__\\s+test_parameters\\s+(.*)/; print "\$\$p-coverage,\$\$c"' tests/\$(TEST)`,\$(COVER_OPTIONS) tests/\$(TEST)

COVER_PARAMETERS = \$(PERL) -e '\$\$l = qx|grep __COVER__ \$\$ARGV[0]|; \$\$u = "-uncoverable_file \$\$1" if \$\$l =~ /__COVER__\\s+uncoverable_file\\s+(.*)/; (\$\$p) = \$\$l =~ /__COVER__\\s+cover_parameters\\s+(.*)/; print "\$\$u \$\$p"' tests/\$(TEST)

html : _run
\t \$(PERL) -Mblib bin/cover `\$(COVER_PARAMETERS)` -report html

basic : _run
\t \$(PERL) -Mblib bin/cover `\$(COVER_PARAMETERS)` -report html_basic

subtle : _run
\t \$(PERL) -Mblib bin/cover `\$(COVER_PARAMETERS)` -report html_subtle

out : _run
\t \$(PERL) -Mblib bin/cover `\$(COVER_PARAMETERS)` -report text > \$(TEST).out

text : out
\t \$(VISUAL) \$(TEST).out

wrun : pure_all
\t \$(PERL) \$(TAINT) -Iblib/lib -Iblib/arch -MDevel::Multithreaded::Cover=-ignore,blib,-merge,0 tests/\$(TEST)

prove : pure_all
\t \$(PERL) -Iutils -MDevel::Multithreaded::Cover::BuildUtils=prove_command -le '\$\$c = prove_command and print \$\$c and system \$\$c'

t : pure_all
\t \$(PERL) -Mblib bin/cover -delete
\t exec make test HARNESS_OPTIONS=j`\$(PERL) -Iutils -MDevel::Multithreaded::Cover::BuildUtils=nice_cpus -e 'print nice_cpus'`:c HARNESS_TIMER=1

no_replace_ops_t : pure_all
\t DEVEL_MULTITHREADED_COVER_OPTIONS=-replace_ops,0 exec make t

DB = cover_db

dump :
\t \$(PERL) -Mblib bin/cover -dump_db \$(DB)

diff : out
\t \$(PERL) utils/makeh strip_criterion 'time' \$(TEST).out
\t \$(PERL) utils/makeh strip_criterion ' pod' \$(TEST).out
\t gold="`\$(PERL) -Mblib -MDevel::Multithreaded::Cover::Test -e '\$\$t = Devel::Multithreaded::Cover::Test->new(qq(\$(TEST))); print join qq(.), \$\$t->cover_gold'`" && \$(EDITOR) -d "\$\$gold" \$(TEST).out

gold : pure_all
\t \$(PERL) utils/create_gold \$(TEST)

all_test :
\t exec \$(PERL) utils/all_versions make t

all_gold :
\t \$(PERL) utils/create_all_gold \$(TEST)

_delete_db : pure_all
\t rm -rf cover_db

_self_cover_tests : @{[sort values %tests]}
\t DEVEL_MULTITHREADED_COVER_SELF=1 \$(PERL) -Mblib -MDevel::Multithreaded::Cover bin/cover -silent -write cover_db @{[sort values %tests]}

self_cover : _self_cover_reports
\t \$(PERL) -Mblib bin/cover -report html_basic -launch
\t \$(PERL) -Mblib bin/cover -report vim

ok :
\t \@$Perlbug -okay  || echo "Please send your report manually to $Author"

nok :
\t \@$Perlbug -nokay || echo "Please send your report manually to $Author"
    ] . "\n" .

    join "\n",
         map("$tests{$_} : _delete_db\n" .
             "\t \@echo Running $tests{$_}\n" .
             "\t \@rm -rf $tests{$_}\n" .
             "\t \@DEVEL_MULTITHREADED_COVER_SELF=1 \$(PERL) -Mblib -MDevel::Multithreaded::Cover=-db,$tests{$_},-silent,1,-coverage,all,-ignore,tests/,-coverage,pod-also_private-xx $_\n",
             sort keys %tests),

         "_self_cover_reports : @{[map qq(report_$_), @reports]}\n",

         map("report_$_ : _self_cover_tests\n" .
             "\t \@echo Generating $_ report\n" .
             "\t \@DEVEL_MULTITHREADED_COVER_SELF=1 \$(PERL) -Mblib -MDevel::Multithreaded::Cover bin/cover -silent -report $_ > /dev/null\n",
             @reports)
}
